cluster-controller/src/main/java/io/strimzi/controller/cluster/resources/TopicController.java

```diff
@@ -0,0 +1,232 @@
+    protected static final String DEFAULT_FULL_RECONCILIATION_INTERVAL = "15 minutes";
+    protected static final String DEFAULT_ZOOKEEPER_SESSION_TIMEOUT = "20 seconds";
+
+    // Configuration keys
+    public static final String KEY_CONFIG = "topic-controller-config";
+
+    // Topic Controller configuration keys
+    public static final String KEY_KAFKA_BOOTSTRAP_SERVERS = "STRIMZI_KAFKA_BOOTSTRAP_SERVERS";
+    public static final String KEY_ZOOKEEPER_CONNECT = "STRIMZI_ZOOKEEPER_CONNECT";
+    public static final String KEY_NAMESPACE = "STRIMZI_NAMESPACE";
+    public static final String KEY_FULL_RECONCILIATION_INTERVAL = "STRIMZI_FULL_RECONCILIATION_INTERVAL";
+    public static final String KEY_ZOOKEEPER_SESSION_TIMEOUT = "STRIMZI_ZOOKEEPER_SESSION_TIMEOUT";
+
+    /**
+     * Constructor
+     *
+     * @param namespace Kubernetes/OpenShift namespace where cluster resources are going to be created
+     * @param cluster   overall cluster name
+     */
+    protected TopicController(String namespace, String cluster) {
+
+        super(namespace, cluster);
+        this.name = topicControllerName(cluster);
+        this.image = DEFAULT_IMAGE;
+        this.replicas = DEFAULT_REPLICAS;
+        this.healthCheckPath = "/";
+        this.healthCheckTimeout = DEFAULT_HEALTHCHECK_TIMEOUT;
+        this.healthCheckInitialDelay = DEFAULT_HEALTHCHECK_DELAY;
+
+        // create a default configuration
+        this.config = new TopicControllerConfig();
+        this.config
+                .withImage(DEFAULT_IMAGE)
+                .withKafkaBootstrapServers(defaultBootstrapServers(cluster))
+                .withZookeeperConnect(defaultZookeeperConnect(cluster))
+                .withNamespace(namespace)
+                .withReconciliationInterval(DEFAULT_FULL_RECONCILIATION_INTERVAL)
+                .withZookeeperSessionTimeout(DEFAULT_ZOOKEEPER_SESSION_TIMEOUT);
+    }
+
+    public static String topicControllerName(String cluster) {
+        return cluster + TopicController.NAME_SUFFIX;
+    }
+
+    protected static String defaultZookeeperConnect(String cluster) {
+        // TODO : exposing the ZookeeperCluster.NAME_SUFFIX as public
+        return cluster + "-zookeeper" + ":" + DEFAULT_ZOOKEEPER_PORT;
+    }
+
+    protected static String defaultBootstrapServers(String cluster) {
+        // TODO : exposing the KafkaCluster.NAME_SUFFIX as public
+        return cluster + "-kafka" + ":" + DEFAULT_BOOTSTRAP_SERVERS_PORT;
+    }
+
+    public void setConfig(TopicControllerConfig config) {
+        this.config = config;
+    }
+
+    public TopicControllerConfig getConfig() {
+        return this.config;
+    }
+
+    /**
+     * Create a Topic Controller from the related ConfigMap resource
+     *
+     * @param cm ConfigMap with topic controller configuration
+     * @return Topic Controller instance, null if not configured in the ConfigMap
+     */
+    public static TopicController fromConfigMap(ConfigMap cm) {
```

Comment:
I wonder if it might be worth changing the method and/or parameter name here (and also for the Zk one) to emphasize that fact that this is the "Kafka Cluster CM", and not a specific Topic controller one.

Reply:
Agree. Maybe changing the name not only for the Zk and Tc but even for the Kafka as well having consistency across all these "resources". Maybe something like `kafkaClusterCm` ?